README.md
# Jarolift_MQTT

Controlling Jarolift(TM) TDEF 433MHz radio shutters via ESP8266 and CC1101 Transceiver Module in asynchronous mode.
Experimental version.
Use at your own risk. For private/educational use only. (Keeloq algorithm licensed only to TI Microcontrollers)
This project is not affiliated in any way with the vendor of the Jarolift components.
Jarolift is a Trademark of Schöneberger Rolladenfabrik GmbH & Co. KG

This is the MQTT version of Jarolift_RX_TX_FHEM, which was originally developed to operate with the 
FHEM home automation server. You find the original FHEM version on the project's homepage.


### Author

Written by Steffen Hille in Nov, 2017


### Project Homepage and Forum

The project home is here: [Project Home](http://www.bastelbudenbuben.de/2017/04/25/protokollanalyse-von-jarolift-tdef-motoren/)
If you need support please use the forum [Project Forum](http://www.bastelbudenbuben.de/forum/)


### License
The main code of Jarolift_MQTT is licensed under the GPLv3. 
The provided libraries have different licenses, look into the respective files for more information.
* CC1101 library        LGPL-License
* DoubleResetDetector   MIT License
* Keeloq library        unknown License
* PubSubClient          MIT License
* simpleDSTadjust       unknown License

### Necessary Hardware:

* a NodeMCU board or similar, based on ESP8266 microcontroller
* a CC1101 Transmitter board
* some connection cables

Photos and hints on project home website

 
### Prepare Arduino IDE and libraries

Download an install the Arduino IDE from [Arduino website](//www.arduino.cc/en/Main/Software)
(as time of writing the current version is 1.8.5)

Locate your arduino sketch directory - this is `$HOME/Arduino/` in linux

Download and / or unpack this project folder to a subfolder of this sketch directory

Locate the user library folder of your arduino sketch directory

Default under linux: `$HOME/Arduino/libraries/`

* copy the folder `DoubleResetDetector` to the arduino library folder
* copy the folder `PubSubClient` to the arduino library folder
* copy the folder `KeeloqLib` to the arduino library folder
* copy the folder `simpleDSTadjust` to the arduino library folder

### when updating from earlier Jarolift_MQTT releases

* in earlier releases, the folder `PubSubClient-2.6.09` was delivered. Make sure you remove `PubSubClient-2.6.09` from the library folder and put the current `PubSubClient` folder there!


### Compile

Open the Arduino IDE and open the project .ino file (`Jarolift_MQTT.ino`)
TODO write a more detailed description

#### Check the board settings
In menu "Tools" set the correct board type. If you have any kind of NodeMCU, the board
``NodeMCU 1.0 (ESP-12E Module`` will probably work.

You may look at the wiki for [pictures of the modules](https://github.com/madmartin/Jarolift_MQTT/wiki/Devices) or at this [comparison of boards](https://frightanic.com/iot/comparison-of-esp8266-nodemcu-development-boards/).

There are two possible pitfalls when choosing the wrong board type or changing between board types:
* the LED on the board may work / not work - necessary for the LED heardbeat and Admin-Mode.
* take a look at the flash size and the SPIFFS setting - a changed SPIFFS setting needs a new upload of the SPIFFS data files and an existing configuration may be not be found on next boot, then it is initialized with default values.

Note on the LED on the board:

The following settings have proofed to work:

##### ESP Core 2.4.1 / NodeMCU v1.0/V2
* Board: "NodeMCU 1.0 (ESP-12E Module)"
* Flash Size: "4M (1M SPIFFS)"
* lwIP variant: "v2 Lower Memory"
* CPU Frequency: "80 MHz"
* Upload speed: "115200"

##### ESP Core 2.4.1 / NodeMCU v3 and WeMos mini & its clones
* Board: "WeMos mini D1 R2 & mini"
* Flash Size: "4M (1M SPIFFS)"
* lwIP variant: "v2 Lower Memory"
* CPU Frequency: "80 MHz"
* Upload speed: "115200"


Dont forget to set the correct serial port (Tools->Port) after plugging the USB connector of the NodeMCU into the computer.

Upload the compiled sketch into the NodeMCU. Second, you need to upload the files from the data directory into the SPIFFS area. This is only necessary on first use of the NodeMCU and then only after the content of the files has changed.


### Uploading files to SPIFFS (ESP8266 internal filesystem)

*ESP8266FS* is a tool which integrates into the Arduino IDE. It adds a
menu item to *Tools* menu for uploading the contents of sketch data
directory into ESP8266 flash file system.

-  Download the [SPIFFS tool](https://github.com/esp8266/arduino-esp8266fs-plugin/releases/download/0.3.0/ESP8266FS-0.3.0.zip)
-  In your Arduino sketchbook directory, create `tools` directory if
   it doesn't exist yet
-  Unpack the tool into `tools` directory (the path will look like
   `$HOME/Arduino/tools/ESP8266FS/tool/esp8266fs.jar`)
-  Restart Arduino IDE
-  Open the Jarolift_MQTT sketch
-  Make sure you have selected a board, port, and closed Serial Monitor
-  Select Tools > ESP8266 Sketch Data Upload. This should start
   uploading the files into ESP8266 flash file system. When done, IDE
   status bar will display `SPIFFS Image Uploaded` message.


### Cabling instructions
ESP | CC1101 | Remark
------|------|------
D8 | CSN
D2 | GDO0 | (perhaps 1Kohm to GND)
D1 | GDO2
D6 | SO(GDO1)
D5 | SCLK
D7 | SI
GND | GND
VCC | VCC | 3.3 Volt !!!


### Setup instructions

The configuration of the Jarolift Dongle is stored in the EEPROM memory of the ESP8266. On first initialisation, when no configuration is found, it is initialized with some default values and the Dongle turns on the Admin-Mode.

In Admin-Mode, the blue LED on the ESP submodule is turned on and the Dongle creates an WLAN-Access-Point with the SSID `Jarolift-Dongle`, protectet with the WPA-Passwort `12345678`. Now you have 180 seconds (3 minutes) time to connect to the WLAN Accesspoint and visit the configuration webserver on

http://192.168.4.1

Admin-Mode quits after the 180 second timeout or when you restart the Dongle from the configuration webserver.

If you need the Admin-Mode later, just press the "Reset" button on the NodeMCU module two times within 10 seconds. This double-reset will be detected and the Dongle enters Admin-Mode again, showing this with the blue LED turned on.

The running Jarolift Dongle does some debug output on the serial console.
Console Speed is 115200 Bit/s


### Known issues

* after flashing the NodeMCU board, on the first run of the fresh flashed sketch, any kind of restart (through WebUI or after a crash) of the ESP8266 may not properly work, causes the NodeMCU board board to hang. On the serial monitor, you see output like
```
handleFileRead: /favicon.png
 [INFO] - Writing Config to EEPROM

 ets Jan  8 2013,rst cause:2, boot mode:(1,6)


 ets Jan  8 2013,rst cause:4, boot mode:(1,6)

wdt reset
```
* this is a hardware issue, solution: power cycle your NodeMCU after flashing, pressing the reset button is not sufficient

* sometimes opening the "system" page in the WebUI causes a crash (Exception 28) of the ESP8266. The cause is not found yet.

### Contribute

You can contribute to Jarolift_MQTT by
- providing Pull Requests (Features, Proof of Concepts, Language files or Fixes)
- testing new released features and report issues
- please try to follow the [Style-Guide](https://www.gnu.org/prep/standards/html_node/Writing-C.html#Writing-C)
  [another Style-Guide](https://google.github.io/styleguide/cppguide.html)


CHANGES.md
## Changelog

**Usage note:** please add new changelog information here at the beginning

##### 2020-04-12
* merged github pull requests #22 and #23
*  #22 delphin0850 committed on 25 Mar 2019: HTML Fix to save the parameters
*  #23 mvn23 committed on 13 Apr 2019: Fix serial generating and parsing
* update library "pubsubclient" to v2.7

##### 2018-08-09
* now it's possible to use the dongle only via the WebUI, by configure NO MQTT broker 
* a heartbeat function lets blink the NodeMCU/ESP's LED every 4sec

##### 2018-07-07 v0.7
* rework use of web_log_message[] array, now a circular buffer
* fix logging of "incoming MQTT command unknown"

##### 2018-07-01
* merge https://github.com/madmartin/Jarolift_MQTT/pull/5/ (partially)
* adds SNTP time fetch and timestamps in log

##### 2018-06-24 v0.6
* make debug output on serial console configurable
* fix Firefox issue in edit channel names (thanks to sidddy)
* add mqtt "sendconfig" command: "cmd/jarolift/sendconfig". the dongle replies with json encoded config data
* check and confirm running with ESP8266 core 2.4.1
* WebUI: enhance layout of system menu, display version
* system menu: serial number prefix is displayed and configurable in hexadecimal now
* system menu: new field: device counter now configurable

##### 2018-06-09 v0.6-rc2
* add logging of WiFi connect and disconnect events. WiFi reconnect works well when compiled with ESP8266 core 2.3.0

##### 2018-06-05 v0.6-rc2
* send device counter to MQTT server after each command
* add subdirectory /doc for misc documentation, beginning with file describing EEPROM memory layout
* implement new connect logic for MQTT, which makes an effective reconnect
* add "uptime in seconds" timestamp to log

##### 2018-06-01 v0.6-rc1
* versionize config data in EEPROM
* introduce configurable mqtt devicetopic
* fix LWT message handling when devicetopic is reconfigured
* Web-UI:
  * explain MQTT better
  * add example for Home Assistant
  * on shutter page: add channelnumber to each channel
* v0.6-rc1
* program version is logged on startup

##### 2018-05-24
* add LWT to mqtt connect
* mqtt client-id must be unique in the network, so add code to append ESP-ID to default
  client-id to make it unique

##### 2018-05-18
* Admin-Mode changed to Access-Point only
* add double reset detector to re-enter Admin-Mode after initial setUpClass
* fixed lots of typos
* fixed inconsistent commenting style
* fixes wording in config webserver

##### 2018-05-15 v0.5
* changed length of SSID to 32 and WPA passwort to 64. needs reconfiguration of MQTT port number
* insert GPLv3 licensing information into all source files

##### 2017-11-11 v0.4a
* improved senden-routine to prevent wdt to reset the device and get more compatibility (controlling and learing).

##### 2017-11-02
* added a switch (bool regular_learn = true; )to choose which learn method to use.
* Regular method is Up+Down at the same time followed by a stop.

##### 2017-10-30
* fixed an error which prevented to learn a motor to the dongle
* reconfigured wifi as STA.

##### 2017-08-30 v0.4
* Added ESP-Reset via fhem. Added shade command to set the shutter to a predefined position.
* Position is set via the "ssp(channel)" cmd.
* Added Group Reception.

##### 2017-08-19 v0.3
* Added WiFi reconnect routine. (THX to Markus)

##### 2017-07-28 v0.2
* Modified rx/tx switch for not getting stuck (line 1182/1192)

##### 2017-07-20 v0.1
* added reception of jarolift remotes and fhem status update



global.h
/*  Controlling Jarolift TDEF 433MHZ radio shutters via ESP8266 and CC1101 Transceiver Module in asynchronous mode.
    Copyright (C) 2017-2018 Steffen Hille et al.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef GLOBAL_H
#define GLOBAL_H

#define PROGRAM_VERSION "v0.7"

#define ACCESS_POINT_NAME      "Jarolift-Dongle"  // default SSID for Admin-Mode
#define ACCESS_POINT_PASSWORD  "12345678"         // default WLAN password for Admin-Mode
#define AdminTimeOut           180                // Defines the time in seconds, when the Admin-Mode will be disabled
#define MQTT_Reconnect_Interval 30000             // try connect to MQTT server very X milliseconds
#define NTP_SERVERS            "0.pool.ntp.org", "1.pool.ntp.org", "2.pool.ntp.org" // List of up to 3 NTP servers
#define TIMEZONE               +1                 // difference localtime to UTC/GMT in hours

struct dstRule StartRule = {"CEST", Last, Sun, Mar, 2, 3600}; // Daylight time
struct dstRule EndRule = {"CET", Last, Sun, Oct, 2, 0};       // Standard time
simpleDSTadjust dstAdjusted(StartRule, EndRule);  // Setup simpleDSTadjust Library rules

ESP8266WebServer server(80);                      // The Webserver
WiFiClient espClient;
PubSubClient mqtt_client(espClient);              // mqtt client instance
long mqttLastConnectAttempt = 0;                  // timepoint of last connect attempt in milliseconds
WiFiEventHandler gotIpEventHandler, disconnectedEventHandler;

boolean AdminEnabled = false;                     // Admin-Mode opens AccessPoint for configuration
int AdminTimeOutCounter = 0;                      // Counter for Disabling the Admin-Mode
boolean wifi_disconnect_log = true;               // Flag to avoid repeated logging of disconnect events
int led_pin = LED_BUILTIN;                        // GPIO Pin number for LED
Ticker tkHeartBeat;                               // Timer for HeartBeat
unsigned short devcnt = 0x0;                      // Initial 16Bit countervalue, stored in EEPROM and incremented once every time a command is send
int cntadr = 110;                                 // EEPROM address where the 16Bit counter is stored.

String web_cmd = "";                              // trigger to run a command whenever a action button has been pressed on the web interface
int web_cmd_channel;                              // keeps the respective channel ID for the web_cmd

#define NUM_WEB_LOG_MESSAGES 42                   // number of messages in the web-UI log page
String web_log_message[NUM_WEB_LOG_MESSAGES];
uint16_t web_log_message_nextfree = 0;
boolean web_log_message_rotated = false;
boolean web_log_message_newline = true;

// boolean settings which will later be configurable through WebUI
boolean debug_mqtt = false;
boolean debug_webui = false;
boolean debug_log_radio_receive_all = false;
boolean mqtt_send_radio_receive_all = true;

struct strConfig {
  uint16_t cfgVersion;
  String  ssid;
  String  password;
  byte    ip[4];
  byte    netmask[4];
  byte    gateway[4];
  boolean dhcp;
  byte    mqtt_broker_addr[4];
  String  mqtt_broker_port;
  String  mqtt_broker_client_id;
  String  mqtt_broker_username;
  String  mqtt_broker_password;
  String  mqtt_devicetopic;
  String  master_msb;
  String  master_lsb;
  unsigned long ulMasterMSB;
  unsigned long ulMasterLSB;
  boolean learn_mode;         // If set to true, regular learn method is used (up+down, followed by stop).
  // If set to false another method for older versions of Jarolift motors is used.
  String  serial;             // starting serial number as string
  uint32_t serial_number;     // starting serial number as integer
  String  channel_name[16];
  //int shutter_time[16]; //this is set in shutter_information
  // temporary values, for web ui, not to store in EEPROM
  String mqtt_devicetopic_new = ""; // needed to figure out if the devicetopic has been changed
  String new_serial = "";
  String new_devicecounter = "";
  boolean set_and_generate_serial = false;
  boolean set_devicecounter = false;
} config;


//####################################################################
// Initalize log message array with empty strings
//####################################################################
void InitLog()
{
  for ( int i = 0; i < NUM_WEB_LOG_MESSAGES; i++ ) {
    web_log_message[i] = "";
  }
} // void InitLog

//####################################################################
// Function to write Log to both, serial and Weblog
//####################################################################
void WriteLog(String msg, boolean new_line = false)
{
  // web_log_message[] is an array of strings, used as a circular buffer
  // web_log_message_nextfree points to the next free-to-use buffer line
  // when web_log_message_nextfree becomes larger than NUM_WEB_LOG_MESSAGES,
  // it is reset to 0 and web_log_message_rotated=true
  // web_log_message_newline indicates whether we are at the beginning of a line or not

  if (web_log_message_newline) {
    char *dstAbbrev;
    time_t now = dstAdjusted.time(&dstAbbrev);
    struct tm * timeinfo = localtime(&now);
    char buffer[30];
    strftime (buffer, 30, "%Y-%m-%d %T", timeinfo);
    web_log_message[web_log_message_nextfree] = (String)buffer + " " + dstAbbrev + " " + msg;
    Serial.print((String)buffer + " " + dstAbbrev + " " + msg);
    web_log_message_newline = false;
  } else {
    web_log_message[web_log_message_nextfree] += " " + msg;
    Serial.print(" " + msg);
  }
  if (new_line == true) {
    web_log_message_nextfree++;
    if (web_log_message_nextfree >= NUM_WEB_LOG_MESSAGES) {
      web_log_message_nextfree = 0;
      web_log_message_rotated = true;
    }
    Serial.println();
    web_log_message_newline = true;
  }
} // void WriteLog

//####################################################################
// Function to connect to Wifi
//####################################################################
void ConfigureWifi()
{
  WriteLog("[INFO] - WiFi connecting to", false);
  WriteLog(config.ssid, true);
  WiFi.mode(WIFI_STA);
  WiFi.begin (config.ssid.c_str(), config.password.c_str());
  if (!config.dhcp)
  {
    WiFi.config(IPAddress(config.ip[0], config.ip[1], config.ip[2], config.ip[3] ),  IPAddress(config.gateway[0], config.gateway[1], config.gateway[2], config.gateway[3] ) , IPAddress(config.netmask[0], config.netmask[1], config.netmask[2], config.netmask[3] ));
  }
  wifi_disconnect_log = true;
} // void ConfigureWifi

//################################################################
// Function to write newest version of configuration into EEPROM
//################################################################
void WriteConfig()
{
  WriteLog("[INFO] - write config to EEPROM", true);
  EEPROM.write(120, 'C');
  EEPROM.write(121, 'f');
  EEPROM.write(122, 'g');
  EEPROM.put(123, config.cfgVersion);

  EEPROM.write(128, config.dhcp);

  EEPROM.write(144, config.ip[0]);
  EEPROM.write(145, config.ip[1]);
  EEPROM.write(146, config.ip[2]);
  EEPROM.write(147, config.ip[3]);

  EEPROM.write(148, config.netmask[0]);
  EEPROM.write(149, config.netmask[1]);
  EEPROM.write(150, config.netmask[2]);
  EEPROM.write(151, config.netmask[3]);

  EEPROM.write(152, config.gateway[0]);
  EEPROM.write(153, config.gateway[1]);
  EEPROM.write(154, config.gateway[2]);
  EEPROM.write(155, config.gateway[3]);

  EEPROM.write(156, config.mqtt_broker_addr[0]);
  EEPROM.write(157, config.mqtt_broker_addr[1]);
  EEPROM.write(158, config.mqtt_broker_addr[2]);
  EEPROM.write(159, config.mqtt_broker_addr[3]);

  WriteStringToEEPROM(164, config.ssid);
  WriteStringToEEPROM(196, config.password);

  WriteStringToEEPROM(262, config.mqtt_broker_port);
  WriteStringToEEPROM(270, config.master_msb);
  WriteStringToEEPROM(330, config.master_lsb);
  EEPROM.write(350, config.learn_mode);
  WriteStringToEEPROM(366, config.serial);

  WriteStringToEEPROM(375, config.mqtt_broker_client_id);
  WriteStringToEEPROM(400, config.mqtt_broker_username);
  WriteStringToEEPROM(450, config.mqtt_broker_password);
  WriteStringToEEPROM(1300, config.mqtt_devicetopic);

  WriteStringToEEPROM(500, config.channel_name[0]);
  WriteStringToEEPROM(550, config.channel_name[1]);
  WriteStringToEEPROM(600, config.channel_name[2]);
  WriteStringToEEPROM(650, config.channel_name[3]);
  WriteStringToEEPROM(700, config.channel_name[4]);
  WriteStringToEEPROM(750, config.channel_name[5]);
  WriteStringToEEPROM(800, config.channel_name[6]);
  WriteStringToEEPROM(850, config.channel_name[7]);
  WriteStringToEEPROM(900, config.channel_name[8]);
  WriteStringToEEPROM(950, config.channel_name[9]);
  WriteStringToEEPROM(1000, config.channel_name[10]);
  WriteStringToEEPROM(1050, config.channel_name[11]);
  WriteStringToEEPROM(1100, config.channel_name[12]);
  WriteStringToEEPROM(1150, config.channel_name[13]);
  WriteStringToEEPROM(1200, config.channel_name[14]);
  WriteStringToEEPROM(1250, config.channel_name[15]);
  
  for ( int i = 0; i < 16; i++ )
	EEPROMWritelong( 1350 + i*4, shutter_information[i].getShutterTime() );

  EEPROM.commit();
  delay(1000);
} // void WriteConfig

//####################################################################
// Function to read configuration from EEPROM
//####################################################################
boolean ReadConfig()
{
  WriteLog("[INFO] - read config from EEPROM . . .", false);
  config.cfgVersion = 0;

  if (EEPROM.read(120) == 'C' && EEPROM.read(121) == 'F'  && EEPROM.read(122) == 'G' )
  {
    WriteLog("config version 1 found", true);
    config.cfgVersion = 1;
  } else
  {
    if (EEPROM.read(120) == 'C' && EEPROM.read(121) == 'f'  && EEPROM.read(122) == 'g' )
    {
      EEPROM.get(123, config.cfgVersion);
      WriteLog("config version " + (String) config.cfgVersion + " found", true);
    }
  }
  if (config.cfgVersion == 0)
  {
    WriteLog(" no configuration found in EEPROM!!!!", true);
    return false;
  }
  if (config.cfgVersion <= 2) // read config parts up to version 2
  {
    config.dhcp = EEPROM.read(128);
    config.ip[0] = EEPROM.read(144);
    config.ip[1] = EEPROM.read(145);
    config.ip[2] = EEPROM.read(146);
    config.ip[3] = EEPROM.read(147);
    config.netmask[0] = EEPROM.read(148);
    config.netmask[1] = EEPROM.read(149);
    config.netmask[2] = EEPROM.read(150);
    config.netmask[3] = EEPROM.read(151);
    config.gateway[0] = EEPROM.read(152);
    config.gateway[1] = EEPROM.read(153);
    config.gateway[2] = EEPROM.read(154);
    config.gateway[3] = EEPROM.read(155);
    config.mqtt_broker_addr[0] = EEPROM.read(156);
    config.mqtt_broker_addr[1] = EEPROM.read(157);
    config.mqtt_broker_addr[2] = EEPROM.read(158);
    config.mqtt_broker_addr[3] = EEPROM.read(159);

    config.ssid = ReadStringFromEEPROM(164, 32);
    config.password = ReadStringFromEEPROM(196, 64);
    config.mqtt_broker_port = ReadStringFromEEPROM(262, 6);
    config.master_msb = ReadStringFromEEPROM(270, 10);
    config.master_lsb = ReadStringFromEEPROM(330, 10);
    config.learn_mode = EEPROM.read(350);
    config.serial = ReadStringFromEEPROM(366, 10);

    config.mqtt_broker_client_id = ReadStringFromEEPROM(375, 25);
    config.mqtt_broker_username = ReadStringFromEEPROM(400, 25);
    config.mqtt_broker_password = ReadStringFromEEPROM(450, 25);

    config.channel_name[0] = ReadStringFromEEPROM(500, 25);
    config.channel_name[1] = ReadStringFromEEPROM(550, 25);
    config.channel_name[2] = ReadStringFromEEPROM(600, 25);
    config.channel_name[3] = ReadStringFromEEPROM(650, 25);
    config.channel_name[4] = ReadStringFromEEPROM(700, 25);
    config.channel_name[5] = ReadStringFromEEPROM(750, 25);
    config.channel_name[6] = ReadStringFromEEPROM(800, 25);
    config.channel_name[7] = ReadStringFromEEPROM(850, 25);
    config.channel_name[8] = ReadStringFromEEPROM(900, 25);
    config.channel_name[9] = ReadStringFromEEPROM(950, 25);
    config.channel_name[10] = ReadStringFromEEPROM(1000, 25);
    config.channel_name[11] = ReadStringFromEEPROM(1050, 25);
    config.channel_name[12] = ReadStringFromEEPROM(1100, 25);
    config.channel_name[13] = ReadStringFromEEPROM(1150, 25);
    config.channel_name[14] = ReadStringFromEEPROM(1200, 25);
    config.channel_name[15] = ReadStringFromEEPROM(1250, 25);
	
  }
  if (config.cfgVersion == 2)
  { // read config parts of version 2
    config.mqtt_devicetopic = ReadStringFromEEPROM(1300, 20);
	
	// actually this is more config version 3... but should work like that as well
	for ( int i = 0; i < 16; i++ )
		shutter_information[i].setShutterTime( EEPROMReadlong( 1350 + i*4 ) );
	
  } else
  { // upgrade config to version 2
    config.mqtt_devicetopic = "jarolift"; // default devicetopic
    config.cfgVersion = 2;
    // clear EEPROM space
    for (int index = 120 ; index < 4096 ; index++) {
      EEPROM.write(index, 0);
    }
    WriteLog("[INFO] - config update to version 2 - mqtt_devicetopic set to default", true);
    WriteConfig();
  }

  // check is config.serial is hexadecimal
  // if necessary, convert decimal to hexadecimal
  if ((config.serial[0] == '0') && (config.serial[1] == 'x')) {
    // config.serial is hex
    // Serial is 28 bits
    // string serial stores only highest 24 bits,
    // add lowest byte with a shift operation for config.serial_number
    config.serial_number = strtol(config.serial.c_str(), NULL, 16) << 4;
  } else {
    // config.serial is NOT hex
    config.serial_number = strtol(config.serial.c_str(), NULL, 10);
    // Serial is 28 bits.
    // string serial stores only highest 24 bits,
    // remove lowest 4 bits with a shift operation
    char serialNumBuffer[11];
    snprintf(serialNumBuffer, 11, "0x%06x", (config.serial_number >> 4));
    config.serial = serialNumBuffer;
    Serial.printf("convert config.serial to hex: %08u = 0x%08x \n", config.serial_number, config.serial_number);
    Serial.println("config.serial: " + config.serial);
    WriteLog("[INFO] - config version 2 - convert serial decimal->hexadecimal", true);
    WriteConfig();
  }

  char charBufSB[11];
  config.master_msb.toCharArray(charBufSB, config.master_msb.length() + 1);
  config.ulMasterMSB = (int)strtol(charBufSB, NULL, 16);
  config.master_lsb.toCharArray(charBufSB, config.master_lsb.length() + 1);
  config.ulMasterLSB = (int)strtol(charBufSB, NULL, 16);

  return true;
} // boolean ReadConfig

//############################################################################
// Function to initialize configuration, either defaults or read from EEPROM
//############################################################################
void InitializeConfigData()
{
  char chipIdString[10];
  sprintf(chipIdString, "-%08x", ESP.getChipId());

  // apply default config if saved configuration not yet exist
  if (!ReadConfig())
  {
    // DEFAULT CONFIG
    config.cfgVersion = 2;
    config.ssid = "MYSSID";
    config.password = "MYPASSWORD";
    config.dhcp = true;
    config.ip[0] = 192; config.ip[1] = 168; config.ip[2] = 1; config.ip[3] = 100;
    config.netmask[0] = 255; config.netmask[1] = 255; config.netmask[2] = 255; config.netmask[3] = 0;
    config.gateway[0] = 192; config.gateway[1] = 168; config.gateway[2] = 1; config.gateway[3] = 1;
    config.mqtt_broker_addr[0] = 0; config.mqtt_broker_addr[1] = 0; config.mqtt_broker_addr[2] = 0; config.mqtt_broker_addr[3] = 0;
    config.mqtt_broker_port = "1883";
    config.mqtt_broker_client_id = "JaroliftDongle";
    config.mqtt_broker_client_id += chipIdString;
    config.mqtt_devicetopic = "jarolift";
    config.master_msb = "0x12345678";
    config.master_lsb = "0x12345678";
    config.learn_mode = true;
    config.serial = "12345600";
    for ( int i = 0; i <= 15; i++ ) {
      config.channel_name[i] = "";
	  shutter_information[i].setShutterTime(21000);
	  
    }
    WriteConfig();
    WriteLog("[INFO] - default config applied", true);
  }

  // check if mqtt client-ID needs migration to new unique ID
  if (config.mqtt_broker_client_id == "JaroliftDongle") {
    config.mqtt_broker_client_id += chipIdString;
    WriteLog("[INFO] - mqtt_broker_client_id changed to " + config.mqtt_broker_client_id, true);
    WriteConfig();
  }
} // void InitializeConfigData

//####################################################################
// Callback function for Ticker
//####################################################################
void Admin_Mode_Timeout()
{
  AdminTimeOutCounter++;
} // void Admin_Mode_Timeout

//####################################################################
// Callback function for LED HeartBeat
//####################################################################
boolean highPulse = true;
#define HEART_BEAT_CYCLE 4                       // HeartBeat cycle in seconds
void HeartBeat()
{
  float pulse_on  = 0.05;                        // LED on for 50 milliseconds in normal mode
  float pulse_off = HEART_BEAT_CYCLE - pulse_on;

  if (WiFi.status() != WL_CONNECTED)
  {
    pulse_on  = HEART_BEAT_CYCLE / 2;            // LED on for 2 seconds, 2 seconds off while waiting for WiFi connect
    pulse_off = pulse_on;
  }

  if (AdminEnabled)
  {
    AdminTimeOutCounter++;
    pulse_off = 0.05;
    pulse_on  = HEART_BEAT_CYCLE - pulse_off;    // LED off for 50 milliseconds in admin mode
  }

  if (highPulse)
    tkHeartBeat.attach(pulse_on, HeartBeat);
  else
    tkHeartBeat.attach(pulse_off, HeartBeat);

  highPulse = !highPulse;
  digitalWrite(led_pin, highPulse);              // toogle LED
} // void HeartBeat

#endif
